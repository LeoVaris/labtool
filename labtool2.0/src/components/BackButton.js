import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import { Button, Icon } from 'semantic-ui-react'

const BackButton = props => (
  <Link to={props.to}>
    <Button compact disabled={!props.enabled} onClick={props.cleanup || (() => {})}>
      <Icon name="angle left" />
      <span>{props.text}</span>
    </Button>
  </Link>
)

BackButton.propTypes = {
  to: PropTypes.string,
  enabled: PropTypes.bool.isRequired,
  text: PropTypes.string.isRequired,
  cleanup: PropTypes.func
}

const presets = {
  modifyCIPage: {
    to: state => (state.selectedInstance.ohid ? `/labtool/ModifyCourseInstancePage/${state.selectedInstance.ohid}` : `/labtool/courses`),
    text: 'Back to course editing'
  },
  coursePage: {
    to: state => (state.selectedInstance.ohid ? `/labtool/courses/${state.selectedInstance.ohid}` : `/labtool/courses`),
    text: 'Back to course page'
  }
}

const mapStateToProps = (state, ownProps) => {
  let to
  try {
    to = presets[ownProps.preset].to(state)
  } catch (e) {
    to = ownProps.to
  }
  let text
  if (presets[ownProps.preset]) {
    text = presets[ownProps.preset].text
  } else {
    text = ownProps.text || 'Back'
  }
  return {
    to,
    text,
    enabled: to !== undefined
  }
}

export default connect(
  mapStateToProps,
  null
)(BackButton)
