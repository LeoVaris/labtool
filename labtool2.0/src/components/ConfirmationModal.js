import React, { useState } from 'react'
import { Modal, Button, Icon } from 'semantic-ui-react'
import PropTypes from 'prop-types'

//used in ModifyCourseInstanceCodeReviews
export const ConfirmationModal = props => {
  const [open, setOpen] = useState(false)

  const toggleOpen = () => setOpen(true)
  const toggleClose = () => setOpen(false)

  return (
    <Modal size="tiny" open={open} trigger={<Icon id="tag" onClick={toggleOpen} name="window close" size="large" color="red" style={{ float: 'right' }} />}>
      <Modal.Content>
        <Modal.Description>
          <p>Do you wish to remove the following code review:</p>
          <p>
            {props.data.User.firsts} {props.data.User.lastname} reviewing {props.getCurrentReviewer(props.selectedDropdown, props.data.id)}
          </p>
        </Modal.Description>
      </Modal.Content>
      <Modal.Actions>
        <Button negative icon="close" labelPosition="right" color="red" content="No" onClick={toggleClose} />
        <Button positive icon="checkmark" labelPosition="right" content="Yes" onClick={props.removeOne(props.data.id)} />
      </Modal.Actions>
    </Modal>
  )
}
ConfirmationModal.propTypes = {
  data: PropTypes.object.isRequired,
  selectedDropdown: PropTypes.number.isRequired,
  getCurrentReviewer: PropTypes.func.isRequired,
  removeOne: PropTypes.func.isRequired
}
export default ConfirmationModal
