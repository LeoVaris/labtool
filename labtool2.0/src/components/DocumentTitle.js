import React from 'react'
import { Helmet } from 'react-helmet'

export default ({ title }) => (
  <Helmet titleTemplate={'%s | Labtool 2.0'} defaultTitle={'Labtool 2.0'}>
    <title>{title}</title>
  </Helmet>
)
