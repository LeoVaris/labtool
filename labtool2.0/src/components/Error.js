import React from 'react'
import PropTypes from 'prop-types'
import { Message, List } from 'semantic-ui-react'

export const LabtoolError = ({ header, errors }) => (
  <Message
    error
    icon="warning sign"
    header={header || 'Error'}
    content={
      <List bulleted>
        {errors.map(error => (
          <List.Item key={error}>{error}</List.Item>
        ))}
      </List>
    }
  />
)
LabtoolError.propTypes = {
  header: PropTypes.string,
  errors: PropTypes.array
}

export default LabtoolError
