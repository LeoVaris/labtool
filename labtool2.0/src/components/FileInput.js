import React from 'react'
import PropTypes from 'prop-types'
import { Button, Modal, Form, Progress, Message } from 'semantic-ui-react'

export default class FileInput extends React.Component {
  state = { open: false, uploading: false, uploadProgress: null, uploadError: null }

  constructor(props) {
    super(props)

    this.fileReader = new FileReader()
    this.fileReader.onloadstart = () => this.setState({ uploading: true, uploadProgress: 0 })
    this.fileReader.onprogress = progress => {
      if (progress.lengthComputable) {
        this.setState({ uploadProgress: 100 * (progress.loaded / progress.total) })
      }
    }
    this.fileReader.onerror = () => this.setState({ uploadError: this.fileReader.error.message, uploadProgress: null })
    this.fileReader.onload = () => {
      this.props.onFileUploaded(this.fileReader.result)
      this.close()
    }
  }

  open = () => this.setState({ open: true })
  close = () => {
    this.fileReader.abort()
    this.setState({ open: false, uploading: false, uploadProgress: null, uploadError: null })
  }

  handleFile = event => {
    if (event.target.files && event.target.files[0]) {
      this.fileReader.readAsArrayBuffer(event.target.files[0])
    }
  }

  render() {
    const { allowedFileTypes, style, ...props } = this.props
    const { open, uploading, uploadProgress, uploadError } = this.state

    return (
      <React.Fragment>
        <Button disabled={open} onClick={this.open} style={style} {...props}>
          Upload
        </Button>

        <Modal onClose={this.close} open={open}>
          <Modal.Header>Select file</Modal.Header>
          <Modal.Description style={{ padding: 15 }}>
            {!uploading && (
              <Form>
                <Form.Input type="file" name="file" accept={allowedFileTypes && allowedFileTypes.join()} onChange={this.handleFile} onClick={e => (e.target.value = null)} />
              </Form>
            )}
            {uploading && uploadProgress !== null && <Progress percent={uploadProgress} indicating />}
            {uploading &&
              uploadError && (
                <Message error>
                  <p>{this.state.uploadError}</p>
                </Message>
              )}
          </Modal.Description>
          <Modal.Actions>
            <Button content="Close" negative onClick={this.close} />
          </Modal.Actions>
        </Modal>
      </React.Fragment>
    )
  }
}

FileInput.propTypes = {
  onFileUploaded: PropTypes.func.isRequired,
  allowedFileTypes: PropTypes.arrayOf(PropTypes.string),
  style: PropTypes.object
}
