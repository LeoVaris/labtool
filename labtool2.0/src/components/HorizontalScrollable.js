import React from 'react'
import { connect } from 'react-redux'

export class HorizontalScrollable extends React.Component {
  constructor(props) {
    super(props)
    this.content = null
    this.scrollbar = null

    this.mainElementReady = element => {
      this.content = element
      this.maybeResizeBar()
    }
    this.scrollBarReady = element => {
      this.scrollbar = element
      this.maybeResizeBar()
    }
  }

  componentDidMount() {
    window.addEventListener('scroll', this.updateSticky)
    window.addEventListener('resize', this.onResize)
    this.onResize()
  }
  componentWillUnmount() {
    window.removeEventListener('scroll', this.updateSticky)
    window.removeEventListener('resize', this.onResize)
  }

  onResize = () => {
    this.maybeResizeBar()
    this.updateSticky()
  }

  maybeResizeBar = () => {
    if (this.content && this.scrollbar && this.content.offsetWidth !== this.scrollbar.style.width) {
      // make scrollable range as wide as table itself, but keep the
      // scroll bar width as the viewable width of the table
      const oldScrollLeft = this.scrollbar.scrollLeft
      this.scrollbar.style.width = this.scrollbar.style.maxWidth = `${this.content.offsetWidth}px`
      this.scrollbar.children[0].style.width = `${this.content.scrollWidth}px`
      this.scrollbar.children[0].style.height = '1px'
      this.scrollbar.scrollLeft = oldScrollLeft
      this.scrollbar.style.overflowX = this.content.scrollWidth > this.content.offsetWidth ? 'scroll' : 'auto'
    }
  }

  updateScrollX = e => {
    // synchronize scroll positions
    const newX = e.target.scrollLeft
    if (this.content) this.content.scrollLeft = newX
    if (this.scrollbar) this.scrollbar.scrollLeft = newX
  }

  updateSticky = () => {
    if (this.content && this.scrollbar) {
      // viewport viewable range of Y: [windowTop, windowBottom[
      const windowTop = document.documentElement.scrollTop
      const windowBottom = windowTop + window.innerHeight

      // top and bottom Y coordinates of table
      const contentTop = this.content.offsetTop
      const contentBottom = contentTop + this.content.clientHeight

      // the table is visible on or above the viewport?
      const pastTableTop = windowBottom >= contentTop
      // the scroll bar is below the viewport?
      const pastTableBottom = windowBottom >= contentBottom

      // we want to display the scroll bar on the screen if
      //   1. the table is visible on or above the viewport (pastTableTop)
      //   2. the scroll bar is *not* below the viewport (!pastTableBottom)
      const makeSticky = pastTableTop && !pastTableBottom
      if (makeSticky) {
        // sticky: fix to bottom of viewport
        this.scrollbar.style.position = 'fixed'
        this.scrollbar.style.bottom = '0px'
      } else {
        // relative: display where it would be otherwise
        this.scrollbar.style.position = 'relative'
        this.scrollbar.style.bottom = 'none'
      }
    }
  }

  render() {
    return (
      <span>
        <div ref={this.mainElementReady} onScroll={this.updateScrollX} style={{ overflowX: 'hidden' }}>
          {this.props.children}
        </div>
        <div ref={this.scrollBarReady} onScroll={this.updateScrollX} style={{ overflowX: 'scroll', bottom: '0', position: 'sticky' }}>
          <div />
        </div>
      </span>
    )
  }
}

export default connect(
  null,
  {}
)(HorizontalScrollable)
