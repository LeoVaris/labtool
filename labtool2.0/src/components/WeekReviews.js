import React, { useEffect } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Icon, Accordion, Card, Comment, Form, Input, Button, Label, Popup } from 'semantic-ui-react'
import { Link } from 'react-router-dom'
import { LabtoolAddComment } from './LabtoolAddComment'
import ReactMarkdown from 'react-markdown'
import { LabtoolComment } from './LabtoolComment'
import { createOneComment } from '../services/comment'
import { getOneCI, coursePageInformation } from '../services/courseInstance'
import { associateTeacherToStudent } from '../services/assistant'
import { addLinkToCodeReview } from '../services/codeReview'
import { sendEmail } from '../services/email'
import { coursePageReset, updateActiveIndex, toggleCodeReview, selectTag, selectTeacher } from '../reducers/coursePageLogicReducer'
import { resetLoading } from '../reducers/loadingReducer'
import useLegacyState from '../hooks/legacyState'

export const WeekReviews = props => {
  const state = useLegacyState({
    openWeeks: {}
  })

  useEffect(() => {
    if (!props.openWeeks && !Object.keys(state.openWeeks).length && !state.openWeeks[props.selectedInstance.currentWeek - 1]) {
      state.openWeeks = { [props.selectedInstance.currentWeek - 1]: true }
    }
  }, [])

  const isTeacher = () => {
    return props.courseData.role === 'teacher'
  }

  const sortCommentsByDate = comments => {
    return comments.sort((a, b) => {
      return new Date(a.createdAt) - new Date(b.createdAt)
    })
  }

  const gradeCodeReview = (reviewNumber, studentInstanceId) => async e => {
    e.preventDefault()
    const data = {
      reviewNumber,
      studentInstanceId: Number(studentInstanceId),
      points: Number(e.target.points.value)
    }
    props.gradeCodeReview(data)
  }

  const renderComment = (isFinalWeek, comment) => {
    /* This hack compares user's name to comment.from and hides the email notification button when they don't match. */
    const userIsCommandSender = comment.from.includes(props.user.user.firsts) && comment.from.includes(props.user.user.lastname)

    return <LabtoolComment key={comment.id} comment={comment} allowNotify={!isFinalWeek && userIsCommandSender} sendCommentEmail={(isTeacher ? sendCommentEmail : sendStudentEmail)(comment.id)} />
  }

  const getMaximumPoints = week => {
    const checklist = props.selectedInstance.checklists.find(checkl => checkl.week === week)
    if (checklist === undefined || checklist.maxPoints === 0 || checklist.maxPoints === null) {
      return props.selectedInstance.weekMaxPoints
    }
    return checklist.maxPoints
  }

  const sendCommentEmail = commentId => async () => {
    props.sendEmail({
      commentId,
      role: 'teacher'
    })
  }

  const sendWeekEmail = weekId => async () => {
    props.sendEmail({
      weekId,
      role: 'teacher'
    })
  }

  const sendStudentEmail = commentId => async () => {
    props.sendEmail({
      commentId,
      role: 'student'
    })
  }

  const handleClickNative = (e, titleProps) => {
    const { index } = titleProps
    const openWeeks = state.openWeeks
    const numberOfOpen = Object.values(openWeeks).filter(x => x).length

    if (numberOfOpen === 0) {
      // if no reviews open, open the clicked tab
      state.openWeeks = { [index]: true }
    } else if (numberOfOpen === 1) {
      // if one open, open the clicked tab and close everything else,
      // or close everything if we clicked the open one
      state.openWeeks = { [index]: !openWeeks[index] }
    } else {
      // if multiple open, simply toggle open/close
      state.openWeeks = { ...openWeeks, [index]: !openWeeks[index] }
    }
  }
  const handleClick = props.handleClickWeek || handleClickNative

  const handleAddingIssueLink = (reviewNumber, studentInstance) => async e => {
    e.preventDefault()
    const data = {
      reviewNumber,
      studentInstanceId: studentInstance,
      linkToReview: e.target.reviewLink.value
    }
    e.target.reviewLink.value = ''
    props.addLinkToCodeReview(data)
  }

  const handleSubmit = async e => {
    e.preventDefault()
    const content = {
      hidden: isTeacher() ? e.target.hidden.checked : false,
      comment: e.target.content.value,
      from: props.user.user.username,
      week: parseInt(e.target.name, 10)
    }
    console.log(content)
    document.getElementById(e.target.name).reset()
    try {
      await props.createOneComment(content)
    } catch (error) {
      console.error(error)
    }
  }

  const renderGradedWeek = (i, week, studentInstance, isFinalWeek, reviewIndex, isWeekOpen) => (
    <Accordion fluid styled id={`review${i}`}>
      <Accordion.Title active={isWeekOpen} index={i + (isFinalWeek && isTeacher() ? 1 : 0)} onClick={handleClick}>
        <Icon name="dropdown" /> {isFinalWeek ? 'Final Review' : `Week ${week.weekNumber}`}, points {week.points} / {getMaximumPoints(week.weekNumber)}
      </Accordion.Title>
      <Accordion.Content active={isWeekOpen}>
        <h3>Review</h3>
        <Card fluid color="yellow">
          <Card.Content>
            <h4>
              {' '}
              Points {week.points} / {getMaximumPoints(week.weekNumber)}{' '}
            </h4>
            <h4> Feedback </h4>
            <ReactMarkdown>{week.feedback}</ReactMarkdown>{' '}
            {isTeacher() && week.instructorNotes ? (
              <div>
                <br />
                <h4>Review notes for instructors </h4>
                <ReactMarkdown>{week.instructorNotes}</ReactMarkdown>{' '}
              </div>
            ) : (
              <span />
            )}
          </Card.Content>
          {isTeacher() && !isFinalWeek ? (
            <Card.Content style={{ paddingBottom: '5px' }}>
              {week.notified ? (
                <Label>
                  Notified <Icon name="check" color="green" />
                </Label>
              ) : (
                <Button type="button" onClick={sendWeekEmail(week.id)} size="small">
                  Send email notification
                </Button>
              )}
            </Card.Content>
          ) : (
            <span />
          )}
        </Card>
        {isTeacher() && (
          <Link to={`/labtool/reviewstudent/${props.selectedInstance.ohid}/${studentInstance}/${reviewIndex}`}>
            <Popup trigger={<Button circular color="orange" size="tiny" icon={{ name: 'edit', color: 'black', size: 'large' }} />} content={isFinalWeek ? 'Edit final review' : 'Edit review'} />
          </Link>
        )}
        {week.comments.length === 0 ? null : <h4> Comments </h4>}
        <Comment.Group>{week ? sortCommentsByDate(week.comments).map(c => renderComment(isFinalWeek, c)) : <h4> No comments </h4>}</Comment.Group>
        <LabtoolAddComment weekId={week.id} commentFieldId={`${props.courseId}:${week.id}`} handleSubmit={handleSubmit} allowHidden={isTeacher()} />
      </Accordion.Content>
    </Accordion>
  )

  const renderUngradedWeek = (i, studentInstance, isFinalWeek, reviewIndex, isWeekOpen) => (
    <Accordion fluid styled id={`review${i}`}>
      <Accordion.Title active={isWeekOpen} index={i} onClick={handleClick}>
        <Icon name="dropdown" /> {isFinalWeek ? 'Final Review' : `Week ${i + 1}`}{' '}
      </Accordion.Title>
      <Accordion.Content active={isWeekOpen}>
        {isTeacher() && (
          <>
            <h3>Review</h3>
            <Link to={`/labtool/reviewstudent/${props.selectedInstance.ohid}/${studentInstance}/${reviewIndex}`}>
              <Popup trigger={<Button circular color="orange" size="tiny" icon={{ name: 'edit', color: 'black', size: 'large' }} />} content={isFinalWeek ? 'Give Final Review' : 'Review week'} />
            </Link>
          </>
        )}
        <h4> Not Graded </h4>
        <h4> No comments </h4>
      </Accordion.Content>
    </Accordion>
  )

  const renderWeek = (i, week, studentInstance, isFinalWeek) => {
    const openWeeks = props.openWeeks || state.openWeeks
    const reviewIndex = isFinalWeek ? props.selectedInstance.weekAmount + 1 : i + 1
    const isWeekOpen = openWeeks[i + (isFinalWeek && isTeacher() ? 1 : 0)]

    return week ? renderGradedWeek(i, week, studentInstance, isFinalWeek, reviewIndex, isWeekOpen) : renderUngradedWeek(i, studentInstance, isFinalWeek, reviewIndex, isWeekOpen)
  }

  // render link of the repo which is assigned to the student to review
  const renderAssignedCodeReviewLink = cr => {
    if (cr.toReview) {
      return (
        <p>
          <strong>Project to review:</strong> {props.courseData.data.find(data => data.id === cr.toReview).projectName} <br />
          <strong>GitHub:</strong>{' '}
          <a href={props.courseData.data.find(data => data.id === cr.toReview).github} target="_blank" rel="noopener noreferrer">
            {props.courseData.data.find(data => data.id === cr.toReview).github}
          </a>
        </p>
      )
    }
    return (
      <p>
        <span>This student is assigned a repo that doesn&#39;t belong to this course to review</span>
        <br />
        <strong>GitHub:</strong>{' '}
        <a href={cr.repoToReview} target="_blank" rel="noopener noreferrer">
          {cr.repoToReview}
        </a>
      </p>
    )
  }

  const renderCodeReviewTeacher = (cr, studentInstance) => (
    <>
      {renderAssignedCodeReviewLink(cr)}
      <strong>Code review:</strong>{' '}
      {cr.linkToReview ? (
        <a href={cr.linkToReview} target="_blank" rel="noopener noreferrer">
          {cr.linkToReview}
        </a>
      ) : (
        'No review linked yet'
      )}
      {cr.points !== null ? <h4>{cr.points} points</h4> : <h4>Not graded yet</h4>}
      <Form onSubmit={gradeCodeReview(cr.reviewNumber, studentInstance)}>
        <label>Points </label>
        <Input name="points" defaultValue={cr.points ? cr.points : ''} type="number" step="0.01" style={{ width: '100px' }} />
        <Input type="submit" value="Grade" />
      </Form>
    </>
  )

  const renderCodeReviewStudent = cr => (
    <>
      <strong>Points: </strong> {cr.points !== null ? cr.points : 'Not graded yet'}
      <br />
      <strong>GitHub: </strong>
      <a href={cr.toReview.github || cr.repoToReview} target="_blank" rel="noopener noreferrer">
        {cr.toReview.github || cr.repoToReview}
      </a>
      <br /> <br />
      {cr.linkToReview ? (
        <div>
          <strong>Your review: </strong>
          <a href={cr.linkToReview} target="_blank" rel="noopener noreferrer">
            {cr.linkToReview}
          </a>
        </div>
      ) : (
        <div />
      )}
      {props.coursePageLogic.showCodeReviews.indexOf(cr.reviewNumber) !== -1 ? (
        <div>
          {cr.linkToReview ? (
            <div />
          ) : (
            <div>
              <strong>Link your review here:</strong> <br />
              <Form onSubmit={handleAddingIssueLink(cr.reviewNumber, props.courseData.data.id)}>
                <Form.Group inline>
                  <Input
                    type="text"
                    name="reviewLink"
                    icon="github"
                    required={true}
                    iconPosition="left"
                    style={{ minWidth: '25em' }}
                    placeholder="https://github.com/account/repo/issues/number"
                    className="form-control1"
                  />
                </Form.Group>
                <Form.Group>
                  <Button compact type="submit" color="blue" style={{ marginLeft: '0.5em' }}>
                    Submit
                  </Button>
                </Form.Group>
              </Form>
            </div>
          )}
        </div>
      ) : (
        <p />
      )}
    </>
  )

  const renderCodeReview = (i, cr, studentInstance) => {
    const openWeeks = props.openWeeks || state.openWeeks
    const doOpen = openWeeks[i] || (!isTeacher() && cr.points === null)

    return (
      <Accordion key={`codereview${i}`} fluid styled id={`review${i - 1}`}>
        {' '}
        <Accordion.Title className="codeReview" active={doOpen} index={i} onClick={handleClick}>
          <Icon name="dropdown" /> Code Review {cr.reviewNumber} {cr.points !== null ? ', points ' + cr.points : ''}
        </Accordion.Title>
        <Accordion.Content active={doOpen}>{isTeacher() ? renderCodeReviewTeacher(cr, studentInstance) : renderCodeReviewStudent(cr)}</Accordion.Content>
      </Accordion>
    )
  }

  const weekMatcher = i => week => week.weekNumber === i + 1
  const weekAmount = props.selectedInstance.weekAmount

  const normalWeeks = []
  for (let i = 0; i < weekAmount; ++i) {
    normalWeeks.push(renderWeek(i, props.student.weeks.find(weekMatcher(i)), props.studentInstance, false))
  }

  const codeReviews = props.student.codeReviews
    .sort((a, b) => {
      return a.reviewNumber - b.reviewNumber
    })
    .map((cr, indx) => renderCodeReview(weekAmount + indx + 1, cr, props.studentInstance))

  const finalReview = []
  if (props.selectedInstance.finalReview) {
    finalReview.push(renderWeek(weekAmount + codeReviews.length, props.student.weeks.find(weekMatcher(weekAmount)), props.studentInstance, true))
  }

  let weeks
  if (isTeacher()) {
    weeks = []
      .concat(normalWeeks)
      .concat(codeReviews)
      .concat(finalReview)
  } else {
    weeks = []
      .concat(normalWeeks)
      .concat(finalReview)
      .concat(codeReviews)

    /* total points */
    weeks.push(
      <Accordion key="total" fluid styled style={{ marginBottom: '2em' }}>
        <Accordion.Title active={true} index="total">
          <Icon name="check" />
          <strong> Total Points: </strong>
          {(
            props.student.weeks
              .map(week => week.points)
              .reduce((a, b) => {
                return a + b
              }, 0) +
            props.student.codeReviews
              .map(cr => cr.points)
              .reduce((a, b) => {
                return a + b
              }, 0)
          )
            .toFixed(2)
            .replace(/[.,]00$/, '')}
        </Accordion.Title>
      </Accordion>
    )
  }

  return weeks
}

WeekReviews.propTypes = {
  courseId: PropTypes.string.isRequired,
  student: PropTypes.object.isRequired,
  studentInstance: PropTypes.string,
  openWeeks: PropTypes.object,
  handleClickWeek: PropTypes.func,

  user: PropTypes.object.isRequired,
  selectedInstance: PropTypes.object.isRequired,
  courseData: PropTypes.object.isRequired,
  coursePageLogic: PropTypes.object.isRequired,
  loading: PropTypes.object.isRequired
}

const mapStateToProps = (state, ownProps) => {
  return {
    user: state.user,
    selectedInstance: state.selectedInstance,
    courseData: state.coursePage,
    coursePageLogic: state.coursePageLogic,
    courseId: ownProps.courseId,
    loading: state.loading
  }
}

const mapDispatchToProps = {
  createOneComment,
  getOneCI,
  coursePageInformation,
  addLinkToCodeReview,
  coursePageReset,
  toggleCodeReview,
  sendEmail,
  updateActiveIndex,
  resetLoading,
  associateTeacherToStudent,
  selectTag,
  selectTeacher
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(WeekReviews)
