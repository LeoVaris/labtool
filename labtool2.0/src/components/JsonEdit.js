import React from 'react'
import PropTypes from 'prop-types'
import { Form, TextArea, Message, Button, Modal } from 'semantic-ui-react'

import FileInput from './FileInput'

export default class JsonEdit extends React.Component {
  state = { open: false, data: null, error: null }

  close = () => this.setState({ open: false })
  open = () => this.setState({ open: true })

  setData = json => {
    try {
      if (!json) {
        this.setState({ error: null })
      }
      this.setState({ data: json })
      JSON.parse(json)
      this.setState({ error: null })
    } catch (error) {
      if (error instanceof SyntaxError) {
        this.setState({ error: `Failed to parse JSON: ${error.message}` })
      }
    }
  }

  onChange = e => this.setData(e.target.value)

  hasValidData = () => !!this.state.data && !this.state.error

  render() {
    const { initialData, style } = this.props
    const { data, open } = this.state

    return (
      <React.Fragment>
        <Button disabled={open} onClick={this.open} style={style}>
          Edit as JSON
        </Button>

        <Modal onClose={this.close} open={open}>
          <Modal.Header>JSON</Modal.Header>
          <Modal.Description style={{ padding: 15 }}>Pressing &quot;Save&quot; saves immediately and overwrites the current checklist.</Modal.Description>
          <Modal.Description style={{ padding: 15 }}>
            {this.state.error && (
              <Message error>
                <p>{this.state.error}</p>
              </Message>
            )}
            <Form>
              <TextArea rows={40} onChange={this.onChange} style={{ fontFamily: 'monospace' }} value={data} defaultValue={JSON.stringify(initialData, null, 4)} />
            </Form>
          </Modal.Description>
          <Modal.Actions>
            <Button
              disabled={!this.hasValidData()}
              content="Download"
              floated="left"
              as="a"
              href={this.hasValidData() ? `data:application/json,${encodeURI(JSON.stringify(JSON.parse(data), null, 4))}` : undefined}
              download="data.json"
              target="_blank"
            />
            <FileInput
              onFileUploaded={data => {
                try {
                  this.setData(JSON.stringify(JSON.parse(new TextDecoder().decode(data)), null, 4))
                } catch (e) {
                  this.setState({ error: `File had invalid data: ${e.message}` })
                }
              }}
              floated="left"
              allowedFileTypes={['application/json']}
            />
            <Button content="Close" negative onClick={this.close} />
            <Button
              content="Save"
              positive
              disabled={!this.hasValidData()}
              onClick={() => {
                this.props.onImport(JSON.parse(this.state.data))
                this.close()
              }}
            />
          </Modal.Actions>
        </Modal>
      </React.Fragment>
    )
  }
}

JsonEdit.propTypes = {
  onImport: PropTypes.func.isRequired,
  initialData: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
  style: PropTypes.object
}
