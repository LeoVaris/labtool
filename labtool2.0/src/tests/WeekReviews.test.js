import React from 'react'
import { WeekReviews } from '../components/WeekReviews'
import { shallow } from 'enzyme'

describe('<WeekReviews /> as student', () => {
  let wrapper

  const coursePage = {
    role: 'student',
    data: {
      id: 10011,
      github: 'http://github.com/tiralabra1',
      projectName: 'Tiran labraprojekti',
      createdAt: '2018-03-26T00:00:00.000Z',
      updatedAt: '2018-03-26T00:00:00.000Z',
      courseInstanceId: 10011,
      userId: 10011,
      teacherInstanceId: 10011,
      weeks: [],
      codeReviews: [
        {
          toReview: {
            github: 'http://github.com/tiralabra2',
            projectName: 'Tiran toinen labraprojekti'
          },
          repoToReview: null,
          reviewNumber: 1,
          points: 2.0
        },
        {
          toReview: {
            github: 'http://github.com/superprojekti',
            projectName: 'Tira super projekti'
          },
          repoToReview: null,
          reviewNumber: 2,
          points: 1.0
        },
        {
          toReview: {
            github: null,
            projectName: null
          },
          repoToReview: 'https://github.com/userName/arbitraryRepo',
          reviewNumber: 3,
          points: null
        }
      ],
      User: {
        id: 10011,
        username: 'tiraopiskelija1',
        email: 'maarit.opiskelija@helsinki.invalid',
        firsts: 'Maarit Mirja',
        lastname: 'Opiskelija',
        studentNumber: '014578343',
        teacher: false,
        sysop: false,
        createdAt: '2018-03-26T00:00:00.000Z',
        updatedAt: '2018-03-26T00:00:00.000Z'
      },
      Tags: []
    }
  }

  const coursePageLogic = {
    showDropdown: '',
    selectedTeacher: '',
    filterByAssistant: 0,
    filterByTag: [],
    showCodeReviews: [2]
  }

  const loading = {
    loading: false,
    loadingHooks: [],
    redirect: false,
    redirectHooks: [],
    redirectFailure: false
  }

  let mockFn = jest.fn()

  beforeEach(() => {
    wrapper = shallow(
      <WeekReviews
        student={coursePage.data}
        getOneCI={mockFn}
        changeCourseField={mockFn}
        courseData={coursePage}
        coursePageInformation={mockFn}
        associateTeacherToStudent={mockFn}
        selectedInstance={coursePage}
        coursePageLogic={coursePageLogic}
        filterByTag={mockFn}
        getAllTags={mockFn}
        courseReset={mockFn}
        loading={loading}
        resetLoading={mockFn}
        courseId={''}
        user={{}}
        createOneComment={mockFn}
        addLinkToCodeReview={mockFn}
        coursePageReset={mockFn}
        toggleCodeReview={mockFn}
        sendEmail={mockFn}
        updateStudentProjectInfo={mockFn}
        selectTag={mockFn}
        selectTeacher={mockFn}
        modifyOneCI={mockFn}
        studentInstance={''}
        gradeCodeReview={mockFn}
        updateActiveIndex={mockFn}
      />
    )
  })

  describe('WeekReviews Component', () => {
    it('is ok', () => {
      true
    })

    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('renders code review cards', () => {
      expect(wrapper.find('WeekReviewCodeReview').length).toEqual(coursePage.data.codeReviews.length)
    })
  })
})
