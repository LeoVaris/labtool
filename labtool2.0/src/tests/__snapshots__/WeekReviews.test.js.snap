// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WeekReviews /> as student WeekReviews Component should render correctly 1`] = `
Array [
  <Accordion
    fluid={true}
    id="reviewNaN"
    key="codereviewNaN"
    styled={true}
  >
     
    <AccordionTitle
      active={false}
      className="codeReview"
      index={NaN}
      onClick={[Function]}
    >
      <Icon
        as="i"
        name="dropdown"
      />
       Code Review 
      1
       
      , points 2
    </AccordionTitle>
    <AccordionContent
      active={false}
    >
      <strong>
        Points: 
      </strong>
       
      2
      <br />
      <strong>
        GitHub: 
      </strong>
      <a
        href="http://github.com/tiralabra2"
        rel="noopener noreferrer"
        target="_blank"
      >
        http://github.com/tiralabra2
      </a>
      <br />
       
      <br />
      <div />
      <p />
    </AccordionContent>
  </Accordion>,
  <Accordion
    fluid={true}
    id="reviewNaN"
    key="codereviewNaN"
    styled={true}
  >
     
    <AccordionTitle
      active={false}
      className="codeReview"
      index={NaN}
      onClick={[Function]}
    >
      <Icon
        as="i"
        name="dropdown"
      />
       Code Review 
      2
       
      , points 1
    </AccordionTitle>
    <AccordionContent
      active={false}
    >
      <strong>
        Points: 
      </strong>
       
      1
      <br />
      <strong>
        GitHub: 
      </strong>
      <a
        href="http://github.com/superprojekti"
        rel="noopener noreferrer"
        target="_blank"
      >
        http://github.com/superprojekti
      </a>
      <br />
       
      <br />
      <div />
      <div>
        <div>
          <strong>
            Link your review here:
          </strong>
           
          <br />
          <Form
            as="form"
            onSubmit={[Function]}
          >
            <FormGroup
              inline={true}
            >
              <Input
                className="form-control1"
                icon="github"
                iconPosition="left"
                name="reviewLink"
                placeholder="https://github.com/account/repo/issues/number"
                required={true}
                style={
                  Object {
                    "minWidth": "25em",
                  }
                }
                type="text"
              />
            </FormGroup>
            <FormGroup>
              <Button
                as="button"
                color="blue"
                compact={true}
                style={
                  Object {
                    "marginLeft": "0.5em",
                  }
                }
                type="submit"
              >
                Submit
              </Button>
            </FormGroup>
          </Form>
        </div>
      </div>
    </AccordionContent>
  </Accordion>,
  <Accordion
    fluid={true}
    id="reviewNaN"
    key="codereviewNaN"
    styled={true}
  >
     
    <AccordionTitle
      active={true}
      className="codeReview"
      index={NaN}
      onClick={[Function]}
    >
      <Icon
        as="i"
        name="dropdown"
      />
       Code Review 
      3
       
    </AccordionTitle>
    <AccordionContent
      active={true}
    >
      <strong>
        Points: 
      </strong>
       
      Not graded yet
      <br />
      <strong>
        GitHub: 
      </strong>
      <a
        href="https://github.com/userName/arbitraryRepo"
        rel="noopener noreferrer"
        target="_blank"
      >
        https://github.com/userName/arbitraryRepo
      </a>
      <br />
       
      <br />
      <div />
      <p />
    </AccordionContent>
  </Accordion>,
  <Accordion
    fluid={true}
    key="total"
    style={
      Object {
        "marginBottom": "2em",
      }
    }
    styled={true}
  >
    <AccordionTitle
      active={true}
      index="total"
    >
      <Icon
        as="i"
        name="check"
      />
      <strong>
         Total Points: 
      </strong>
      3
    </AccordionTitle>
  </Accordion>,
]
`;
